### PR TITLE
Multi-query attention

### PR DESCRIPTION
**Pull request by jlamypoirier**

### CODE REVIEW

**Review by bigximik — COMMENTED** [review]
cool, maybe just to call it kv_dim instead of k_dim?

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 1]
```diff
@@ -163,14 +163,13 @@ def __init__(self, config, is_cross_attention=False, layer_idx=None):
...
-            if self.attention_type != AttentionType.MULTI_HEAD:
-                self.c_attn = Conv1D((self.num_heads + 2) * self.head_dim, self.embed_dim)
-            elif self.attention_type != AttentionType.MULTI_QUERY_2:
+            if self.attention_type == AttentionType.MULTI_QUERY_2:
                 self.q_attn = Conv1D(self.embed_dim, self.embed_dim)
                 # Keys and values are shared across heads
                 self.kv_attn = Conv1D(2 * self.head_dim, self.embed_dim)
             else:
-                self.c_attn = Conv1D(3 * self.embed_dim, self.embed_dim)
+                k_dim=self.embed_dim if self.attention_type == AttentionType.MULTI_HEAD else self.head_dim
```
Comment (bigximik) [inline src/transformers/models/gpt2/modeling_gpt2.py:171]: cool, maybe just to call it kv_dim instead of k_dim?

**Comment by bigximik** [comment]
@jlamypoirier, I can merge this PR in the morning with anything you will add today and also to merge @lvwerra branch from Hugging Face repo and to move all to gpt2mqa model from here? Will it be OK?

**Comment by jlamypoirier** [comment]
> @jlamypoirier, I can merge this PR in the morning with anything you will add today and also to merge @lvwerra branch from Hugging Face repo and to move all to gpt2mqa model from here? Will it be OK?

I think we should wait a bit, there seems to be some bugs still.

**Comment by jlamypoirier** [comment]
Merged to speed things up, it doesn't affect the rest of transformers anyway

**Pull request merged**